### PR TITLE
Add an aria-live announcer to the loading mode of Button for a11y

### PR DESCRIPTION
**Pull request by jorycunningham**
### Description

Resolves AWSUI-18905

This adds an aria-live announcement that fires every 4 seconds to announce the loading state of the loading button to ensure there is an accessible equivalent to the visual treatment.

To enable this I have added an optional aria-live announcer to the Spinner component and enabled it in the "loading" mode of the button component.

#### Overview of changes:
* Added an `I18nStrings.loadingAriaLiveText` to Spinner to receive a text string to be announced and toggle on the aria-live announcement
* Added  `I18nStrings.loadingAriaLiveText` to the Button component so that it can pass it down to the spinner
* Updated Button/icon-helper to facilitate the passing of the `I18nStrings.loadingAriaLiveText` prop
* Added an aria-live region to the spinner that announces the `1I18nStrings.loadingAriaLiveText` every 4 seconds while the button is in loading mode
* Added loadingAriaLiveText string to button permutations and tab navigation pages
* Updated the LiveRegion Component to render a span rather than a div for the outside container so that it can be used inside a button element. (cc: @timogasda )

### How has this been tested?

Tested in JAWS, NVDA, and VoiceOver.
Ran all tests including Visual regression

[_How can reviewers test these changes efficiently?_]
The Button/ Tab Panel and Button / Permutations pages have this feature enabled on the loading buttons

[_Check for unexpected visual regressions, see [`CONTRIBUTING.md`](CONTRIBUTING.md#run-visual-regression-tests) for details._]

### Documentation changes

[*Do the changes include any API documentation changes?*]
- [X] _Yes, this change contains documentation changes._
- [ ] _No._

### Related Links

Resolves AWSUI-18905

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- [ ] _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](CONTRIBUTING.md#public-apis)._
- [ ] _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](CONTRIBUTING.md#browsers-support)._
- [ ] _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- [ ] _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- [ ] _Changes are covered with new/existing unit tests?_
- [ ] _Changes are covered with new/existing integration tests?_
</details>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by timogasda — CHANGES_REQUESTED** [review]
Left some suggestions and general understanding questions.

It also looks like there are some merge conflicts and the documenter snapshot file is gone completely.

**Comment by timogasda** [comment]
Can you please change the title of the PR to something more tangible? The ticket ID isn't super helpful here.

File: src/button/internal.tsx [hunk 1]
```diff
@@ -102,6 +103,7 @@ export const InternalButton = React.forwardRef(
       variant,
       iconClass: __iconClass,
       iconSize: variant === 'modal-dismiss' ? 'medium' : 'normal',
+      i18nStrings,
```
[the remark this answers is not in the document]
Comment (jorycunningham) [reply src/button/internal.tsx]: Agreed. I figured there would be other uses for this. I'll open a new feature ticket.

File: src/spinner/internal.tsx [hunk 2]
```diff
@@ -1,28 +1,52 @@
...
+    if (loadingText) {
+      let nextText = loadingText;
+      setliveRegionText(nextText);
+
+      const toggleAnnouncer = () => {
+        nextText = nextText === loadingText ? '' : loadingText;
+        setliveRegionText(nextText);
+      };
+
+      const announceTimer = setInterval(toggleAnnouncer, 2000);
```
[the remark this answers is not in the document]
Comment (jorycunningham) [reply src/spinner/internal.tsx]: The 2 second timer toggles the aria-live region from the loading alt text to an empty string, so the net effect is that it announces every 4 seconds while the other updates (the ones that clear it) are there to make sure the region detects a change when the text is added back in.

I moved the 4000ms to a const and annotated it with a comment,

File: src/button/interfaces.ts [hunk 3]
```diff
@@ -136,4 +142,8 @@ export namespace ButtonProps {
      */
     focus(options?: FocusOptions): void;
   }
+
+  export interface I18nStrings {
+    loadingAriaLiveText?: string;
```
[the remark this answers is not in the document]
Comment (jorycunningham) [reply src/button/interfaces.ts]: That's a really good point. I'm going to pull this out into its own prop, we'll likely need flexibility here, for example if there are multiple loading buttons (😨 ) on a page. I think the context will matter here for the recommended text.

**Comment by pan-kot** [comment]
Why would we add this announcer to the actual component(s)? It doesn't rely on any internal state and can be implemented on the customer's end. We can consider adding some helper to encapsulate the announcement interval and the aria-live implementation itself. Compare:

```
<Button loading={loading} loadingAriaLiveText="loading">
  {loading ? 'Loading...' : 'Click me'}
</Button>
``` 

and

```
<>
  <Button loading={loading}>
    {loading ? 'Loading...' : 'Click me'}
  </Button>
  
  // Can be extended easily e.g. if we decide to parametrise the announcement rate.
  // Can be reused for any component that has loading state.
  {loading && <IntervalLiveRegion>loading</IntervalLiveRegion>}
</>
```

**Review by pan-kot — CHANGES_REQUESTED** [review]
Added a question on whether an alternative approach can be used.

File: src/spinner/internal.tsx [hunk 4]
```diff
@@ -1,28 +1,55 @@
...
   __internalRootRef,
   ...props
 }: InternalSpinnerProps) {
   const baseProps = getBaseProps(props);
 
+  const [liveRegionText, setliveRegionText] = useState('');
+  /* Results in the message repeating at twice the value of messageToggleTiming.
+   * In this case 4 seconds was the interval that "felt right" to the a11y team
+   */
+  const messageToggleTiming = 2000;
```
Comment (pan-kot) [inline src/spinner/internal.tsx:26]: Is it 2 seconds or 4 seconds?

Comment (jorycunningham) [reply src/spinner/internal.tsx:26]: It is 4 seconds. It changes state every 2 seconds but only only one state announces, the other clears it. I'll clarify this,

**Comment by jorycunningham** [comment]
> Why would we add this announcer to the actual component(s)? It doesn't rely on any internal state and can be implemented on the customer's end. We can consider adding some helper to encapsulate the announcement interval and the aria-live implementation itself. Compare:
> 
> ```
> <Button loading={loading} loadingAriaLiveText="loading">
>   {loading ? 'Loading...' : 'Click me'}
> </Button>
> ```
> 
> and
> 
> ```
> <>
>   <Button loading={loading}>
>     {loading ? 'Loading...' : 'Click me'}
>   </Button>
>   
>   // Can be extended easily e.g. if we decide to parametrise the announcement rate.
>   // Can be reused for any component that has loading state.
>   {loading && <IntervalLiveRegion>loading</IntervalLiveRegion>}
> </>
> ```
> 
> Another consideration is when there are multiple loading indicators being active at once. Ideally is to avoid it, but I'd imagine that there can be cases like that. When the loading announcer is a separate component -- such situations can be handled easily.

I did it on the spinner so it can be used everywhere the spinner is used, not just a button.as will likely be necessary. We can easily "parameterise" the time on the spinner-level if at some point that becomes necessary, we discussed this in the a11y team and decided there was not a need for it at this time

File: src/spinner/internal.tsx (hunk 4 above)
Comment (jorycunningham) [reply src/spinner/internal.tsx:26]: See: https://github.com/cloudscape-design/components/pull/316#discussion_r984650242

**Pull request closed without merging**